safe/converter.py

```diff
@@ -327,9 +337,15 @@ def encoder(
             attach_regexp = re.compile(r"(" + re.escape(attach) + r")")
             scaffold_str = attach_regexp.sub(val, scaffold_str)
             starting_num += 1
-        # now we need to remove all the parenthesis around difig only number
+        # now we need to remove all the parenthesis around digit only number
         wrong_attach = re.compile(r"\(([\%\d]*)\)")
-        return wrong_attach.sub(r"\g<1>", scaffold_str)
+        scaffold_str = wrong_attach.sub(r"\g<1>", scaffold_str)
```

Comment:
I understood the `r'\1\d` based substitution in the thread in the issue, but my regex is not good enough, could you help me understand why this is changed into `\g<1>\g<2>` here?

Reply:
The syntax `\g<1>\g<2>` is the more modern and clearer alternative to using `r'\1\2'`.  It avoids ambiguity in cases where group numbers might be misinterpreted. For instance,  with` r'\11\2'`,  it's unclear whether we are referring to **group 11** followed by **group 2**, or **group 1** followed by the literal character **'1'** and then **group 2**. 
In contrast, using `\g<11>\g<2>` helps avoid confusion and makes the substitution clear.

Reply:
https://docs.python.org/3/howto/regex.html#search-and-replace

Reply:
thank you, makes sense. 
not to do with the PR but may I also ask why is this happening?
```
wrong_attach = re.compile(r"\(([\%\d]*)\)")
scaffold_str = wrong_attach.sub(r"\g<1>", scaffold_str)
```
 my understanding is this just adds a pair of brackets at the start and end of the scaffold_str?

Reply:
It replaces the  patterns like `(1)` by `1`, since the former is not valid in SMILES. This is similar to the issue you raise. 

In theory, all of those regex subs are to cover the case where we have strange parenthesis during processing. For example for this SMILES: `CCCc1ccccc1`

```
# this is a valid form of the SMILES
c13ccccc1.CCC3
# this is an invalid SMILES that we need to convert to c13ccccc1.CCC3 
c1(3)ccccc1.CCC3
#  this is what we have just fixed by converting to c13ccccc1.CCC3
# It is technically valid SMILES, just not for RDKIT 
c1(-3)ccccc1.CCC3
```

Reply:
Thanks so much for the explanation